misc/package/Data Files/MWSE/core/lib/logger/init.lua

```diff
@@ -0,0 +1,658 @@
+- `writeToFile`: either a boolean saying we should write to a file, or the name of a file to write to. If false (the default), then log messages will be written to `MWSE.log`
+]]
+---@param params mwseLogger.newParams|string?
+---@return mwseLogger
+function Logger.new(params)
+    -- if it's just a string, treat it as the `modName`
+    if not params then
+        params = {}
+    elseif type(params) == "string" then
+        params = {modName=params} 
+    end
+    local modName, moduleName = params.modName, params.moduleName
+    -- do some error checking to make sure `params` are correct
+    if not modName then
+        modName, moduleName = getModInfoFromSource()
+    end
+    assert(modName ~= nil, "Error: Could not create a Logger because modName was nil.")
+    assert(type(modName) == "string", "Error: Could not create a Logger. modName must be a string.")
+
+    if moduleName == nil then
+        modName, moduleName = getModAndModuleNames(modName)
+    end
+
+    local log ---@type mwseLogger?
+
+    -- first try to get it
+    log = Logger.get(modName, moduleName)
+
+    if log then return log end
+
+    -- now we know the log doesn't exist
+
+    log = {
+        modName = modName,
+        moduleName = moduleName,
+        useColors=params.useColors or false,
+        writeToFile = params.writeToFile or false,
+        level = Logger.LEVEL.INFO, -- we'll set it with the dedicated function so we can do fancy stuff
+        includeTimestamp = params.includeTimestamp or false,
+    }
+
+    
+    -- if there are already loggers with this `modName`, get the most recent one, and then
+    -- update this new loggers values to those of the most recent logger. (if those values weren't specified in `params`)
+    if loggers[modName] and #loggers[modName] > 0 then
+        local parent = loggers[modName][#loggers[modName]]
+        for _, k in ipairs{"useColors", "writeToFile", "includeTimestamp"} do
+            if params[k] == nil then
+                log[k] = parent[k]
+            end
+        end
+        -- set the default value to the parent log level, we will call `setLevel` later
+        -- this will do the the following:
+        -- if `params.level` was invalid, nothing will happen and `parent.level` will be used.
+        -- if `params.level` was valid, then every logger registered to this mod will be updated.
+        log.level = parent.level
+    else
+        -- this is the first logger with this `modName`, so we should intialize the array
+        loggers[modName] = {}
+    end
+
+    table.insert(loggers[modName], log)
+
+    setmetatable(log, log_metatable)
+    log:setLevel(params.level)
+
+    if params.writeToFile == nil then
+        log:setWriteToFile(log.writeToFile, true)
+    else
+        log:setWriteToFile(params.writeToFile)
+    end
+
+
+    return log
+end
+
+
+
+---@param writeToFile string|boolean
+---@param onlyThisLogger boolean? should we only update this logger? Default: false
```

Comment:
What do you think of renaming this argument to something like `updateChildren`?

Reply:
i think this would be better. i'll make this change later today